multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/stream/LazyArchiveInputStream.java

```diff
@@ -0,0 +1,106 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LazyArchiveInputStream.class);
+    private static final int BUFFERED_SIZE = 16 * 1024;
+
+    private final FileService fileService;
+    private final List<FileEntry> archiveFileEntries;
+    private final StepLogger stepLogger;
+    private final int archiveSize;
+    private final AtomicInteger totalBytesRead;
+    private final AtomicInteger partIndex;
+
+    private InputStream currentInputStream;
+
+    public LazyArchiveInputStream(FileService fileService, List<FileEntry> archiveFileEntries, StepLogger stepLogger, int archiveSize) {
+        this.fileService = fileService;
+        this.archiveFileEntries = archiveFileEntries;
+        this.stepLogger = stepLogger;
+        this.archiveSize = archiveSize;
+        this.totalBytesRead = new AtomicInteger(0);
+        this.partIndex = new AtomicInteger(0);
+    }
+
+    @Override
+    public synchronized int read() throws IOException {
+        if (currentInputStream == null) {
+            currentInputStream = openBufferedInputStream(archiveFileEntries.get(partIndex.get()));
+        }
+        int c = currentInputStream.read();
+        if (c == -1 && partIndex.get() < archiveFileEntries.size() - 1) {
+            IOUtils.closeQuietly(currentInputStream, e -> LOGGER.warn(e.getMessage(), e));
+            LOGGER.info(MessageFormat.format(Messages.CLOSING_STREAM_FOR_PART_0, partIndex));
+            currentInputStream = openBufferedInputStream(archiveFileEntries.get(partIndex.incrementAndGet()));
+            c = currentInputStream.read();
+        } else if (c == -1) {
+            LOGGER.info(MessageFormat.format(Messages.CLOSING_STREAM_FOR_PART_STREAM_FINISHED_0, partIndex));
+            IOUtils.closeQuietly(currentInputStream, e -> LOGGER.warn(e.getMessage(), e));
+        }
+        if (c >= 0) {
+            totalBytesRead.incrementAndGet();
+        }
+        return c;
+    }
+
+    @Override
+    public synchronized int read(byte[] b, int off, int len) throws IOException {
+        if (currentInputStream == null) {
+            currentInputStream = openBufferedInputStream(archiveFileEntries.get(partIndex.get()));
+        }
+        int bytesRead = currentInputStream.read(b, off, len);
+        if (bytesRead == -1 && partIndex.get() < archiveFileEntries.size() - 1) {
+            IOUtils.closeQuietly(currentInputStream, e -> LOGGER.warn(e.getMessage(), e));
+            LOGGER.info(MessageFormat.format(Messages.CLOSING_STREAM_FOR_PART_0, partIndex));
+            currentInputStream = openBufferedInputStream(archiveFileEntries.get(partIndex.incrementAndGet()));
+            bytesRead = currentInputStream.read(b, off, len);
+        } else if (bytesRead == -1) {
+            LOGGER.info(MessageFormat.format(Messages.CLOSING_STREAM_FOR_PART_STREAM_FINISHED_0, partIndex));
+            IOUtils.closeQuietly(currentInputStream, e -> LOGGER.warn(e.getMessage(), e));
+        }
+        if (bytesRead > 0) {
```

Comment:
maybe use as 'else if'

Reply:
with else if, when bytesRead = currentInputStream.read(b, off, len); returns something totalBytesRead won't be updated.
there are cases where we update bytesRead in the if-else blocks